File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>
using namespace std;

bool comp(pair<int, string> a, pair<int, string> b) {
    return a.first > b.first;
}
bool comp2(pair<int,int> a, pair<int, int> b) {
    if (a.first == b.first) return a.second < b.second;
    return a.first > b.first;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    map<string, int> m; // genre, play
    map<string, vector<pair<int, int>>> list; // genre, play, id
    vector<pair<int, string>> v; // play, genre
    for (int i = 0; i < genres.size(); i++) {
        string genre = genres[i];
        int play = plays[i];
        m[genre] += play;
        list[genre].push_back(make_pair(play, i));
    }
    
    for (auto iter = m.begin(); iter != m.end(); iter++) {
        v.push_back(make_pair(iter->second, iter->first));
        sort(list[iter->first].begin(), list[iter->first].end(), comp2);
    }
    sort(v.begin(), v.end(), comp);

    for (int i = 0; i < v.size(); i++) {
        string genre = v[i].second;
        int k = 0;
        for (auto iter = list[genre].begin(); iter != list[genre].end(); iter++) {
            if (k == 2) break;
            answer.push_back(iter->second);
            k++;
        }
    }

    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    static class Song {
        int id;
        int play;
        
        Song(int id, int play) {
            this.id = id;
            this.play = play;
        }
    }

    public int[] solution(String[] genres, int[] plays) {
        HashMap<String, Integer> genreTotal = new HashMap<>();
        HashMap<String, ArrayList<Song>> genreSongs = new HashMap<>();
        
        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];
            
            genreTotal.put(genre, genreTotal.getOrDefault(genre, 0) + plays[i]);
            
            genreSongs.putIfAbsent(genre, new ArrayList<>());
            genreSongs.get(genre).add(new Song(i, plays[i]));
        }
        
        ArrayList<String> genreList = new ArrayList<>(genreTotal.keySet());
        
        genreList.sort((a, b) -> genreTotal.get(b) - genreTotal.get(a));
        
        ArrayList<Integer> answer = new ArrayList<>();
        
        for (String genre : genreList) {
            ArrayList<Song> songs = genreSongs.get(genre);
            
            songs.sort((a, b) -> {
                if (b.play == a.play) {
                    return a.id - b.id;
                }
                return b.play - a.play;
            });
            
            int count = Math.min(2, songs.size());
            
            for (int i = 0; i < count; i++) {
                answer.add(songs.get(i).id);
            }
        }
        
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}